sale_weekday_delivery/models/sale_order.py

```diff
@@ -0,0 +1,116 @@
+
+    @api.onchange("commitment_date")
+    def _onchange_commitment_date(self):
+        """Warns if commitment date is not a preferred weekday for delivery"""
+        res = super()._onchange_commitment_date()
+        if res:
+            return res
+        if (
+            self.commitment_date
+            and self.partner_shipping_id.delivery_schedule == "fix_weekdays"
+        ):
+            raw_commitment_date_weekday = self.commitment_date.weekday()
+            commitment_date_weekday = WEEKDAY_MAPPING.get(
+                str(raw_commitment_date_weekday)
+            )
+            ps = self.partner_shipping_id
+            if not ps.is_preferred_delivery_weekday(commitment_date_weekday):
+                return {
+                    "warning": {
+                        "title": _(
+                            "Commitment date does not match shipping "
+                            "partner's Delivery schedule preference."
+                        ),
+                        "message": _(
+                            "The delivery date is on a %s, but the shipping "
+                            "partner is set to prefer deliveries on following "
+                            "weekdays:\n%s"
+                            % (
+                                commitment_date_weekday,
+                                '\n'.join(
+                                    [
+                                        "  * %s" % day
+                                        for day
+                                        in
+                                        ps.get_delivery_schedule_preferred_weekdays()]
+                                ),
+                            )
+                        ),
+                    }
+                }
+
+
+class SaleOrderLine(models.Model):
+
+    _inherit = "sale.order.line"
+
+    def _expected_date(self):
+        """Postpone expected_date to next preferred weekday"""
+        expected_date = super()._expected_date()
+        ops = self.order_id.partner_shipping_id
+        if ops.delivery_schedule == "direct":
+            return expected_date
+        return ops._next_delivery_schedule_preferred_date(
+            expected_date
+        )
+
+    def _prepare_procurement_values(self, group_id=False):
+        """Consider delivery_schedule in procurement"""
+        res = super()._prepare_procurement_values(group_id=group_id)
+        if (
+            self.order_id.partner_shipping_id.delivery_schedule != "fix_weekdays"
+            # if a commitment_date is set we don't change the result as lead
+            # time and delivery week days must have been considered
+            or self.order_id.commitment_date
+        ):
+            return res
+        # If no commitment date is set, we must consider next preferred delivery
+        #  weekday to postpone date_planned
+        date_planned = fields.Datetime.to_datetime(res.get("date_planned"))
+        ops = self.order_id.partner_shipping_id
```

Comment:
ops? `partner` sounds better

Reply:
That's what happens when you have to fight with black formatting....